hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/JobRepository.java

```diff
@@ -204,25 +212,51 @@ long uploadJobResources(JobConfig jobConfig) {
     }
 
     private long newJobId() {
-       return idGenerator.newId();
+        return idGenerator.newId();
     }
 
     /**
-     * Unzips the Jar archive and processes individual entries using
-     * {@link #readStreamAndPutCompressedToMap(String, Map, InputStream)}.
+     * Unzips the zip archive and processes individual entries and jar files
      */
-    private void loadJar(Map<String, byte[]> map, URL url) throws IOException {
-        try (JarInputStream jis = new JarInputStream(new BufferedInputStream(url.openStream()))) {
-            JarEntry jarEntry;
-            while ((jarEntry = jis.getNextJarEntry()) != null) {
-                if (jarEntry.isDirectory()) {
+    private void loadZip(Map<String, byte[]> map, URL url) throws IOException {
+        try (ZipInputStream zis = new ZipInputStream(new BufferedInputStream(url.openStream()))) {
+            ZipEntry zipEntry;
+            while ((zipEntry = zis.getNextEntry()) != null) {
+                if (zipEntry.isDirectory()) {
                     continue;
                 }
-                readStreamAndPutCompressedToMap(jarEntry.getName(), map, jis);
+                if (zipEntry.getName().endsWith(".jar")) {
+                    loadJarFromInputStream(map, zis);
+                }
+                readStreamAndPutCompressedToMap(zipEntry.getName(), map, zis);
```

Comment:
`else` should be here, otherwise we'll add resource to classpath with the jar file, but it will be empty because the `zis` is already read. But I would consider removing it altogether. We should only add the jar files from the zip, not all files. The packages of java products typically contains samples, javadocs, licenses, which are not needed and even should not be on the classpath, they take precious space in IMap at best.